internal/cache/store/queue.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"hash/fnv"
+	"sync"
+)
+
+// ShardedUniqueQueue is a queue of write requests
+// for objects. It compacts consecutive create and update
+// requests, provides a slice of channels for consumers.
+// Requests consumed from these channels are mutually
+// exclusive
+type ShardedUniqueQueue interface {
+	AddIfAbsent(Request)
+	GetConsumers() []<-chan func() Request
+}
+
+// NewShardedUniqueQueue creates a sharded queue of write requests
+func NewShardedUniqueQueue(buckets int) ShardedUniqueQueue {
+	queues := make([]chan func() Request, 0, buckets)
+	for i := 0; i < buckets; i++ {
+		queues = append(queues, make(chan func() Request, 100))
+	}
+	return &shardedUniqueQueue{
+		queues:   queues,
+		inflight: make(map[Key]bool),
+	}
+}
+
+type shardedUniqueQueue struct {
+	queues   []chan func() Request
+	inflight map[Key]bool
+	lock     sync.Mutex
+}
+
+// AddIfAbsent puts a request to the queue if it is absent,
+// or if it is a delete request. Deletes are not compacted
+// to prior update or create requests, so all input objects
+// can be created
+func (q *shardedUniqueQueue) AddIfAbsent(r Request) {
```

Comment:
this code is interesting and weird! I have many questions, let's discuss.

Reply:
The idea is to add a write request only if there are no other write requests in the queue. I have gone with excluding delete requests from this, because I want each object to appear in etcd at one point. Example would be a create call and a delete call while the create call was in flight, if you compact the two you can get away with doing nothing, but etcd will be unaware of that object has existed in our cache at one point. So mainly I want this for increased visibility